### DISCUSSION TITLE
Do we need a delay after connecting before sending modbus messages?

### DISCUSSION THREAD

**Discussion by canton7** [nathanmarlor/foxess_modbus#132] [Q&A] [answered]
I'm seeing some some messages in the Facebook group saying that people connecting a lan cable directly to the inverter with StealthChesnut's integration need `delay: 1` in their config, to avoid random inverter crashes.

Do we need to emulate this?

On a related note, my W610 seems to always fail the first request after a connection, so perhaps that's related.

**Comment by paultweedy77** [comment]
I've had this running directly from inverter LAN port for a couple of days and have had a couple of inverter crashes/restarts so far. 
Oddly enough they've coincided with me either mucking around with settings in HA.

In line with that advice from the Facebook group, I've changed in modbus_client.py:

`self._poll_delay = 30 / 1000 if self._config_type == SERIAL else 0`
to
`self._poll_delay = 30 / 1000 if self._config_type == SERIAL else 1`

which will do a 1s delay before each TCP modbus poll - will see if this results in any greater stability.

**Reply by canton7** [reply]
Interesting, let me know what you find!

Note that the `_poll_delay` is applied after *every* poll, whereas the `delay: 1` delay is applied *before* the first poll after a connection, so you might well find it makes no difference.

You might need to put a sleep somewhere in here: https://github.com/nathanmarlor/foxess_modbus/blob/d8cff21624db5084644d5eb02f86500e5c3d0921/custom_components/foxess_modbus/modbus_client.py#L34

**Reply by paultweedy77** [reply]
Good catch, I hadn't read through the code very thoroughly!
I've added 
`time.sleep(1)`
before that return.. we'll see.

**Reply by canton7** [reply]
Although, connect is called all the freaking time (twice per read, regardless of whether you're connected or not), so that will shoot performance in the foot. Probably delaying only if it wasn't connected and now it is (ie the condition which calls setsockopt above)

**Reply by paultweedy77** [reply]
```
        if not was_connected and is_connected:
            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, True)
            time.sleep(1)
        return is_connected

```
I'm not great at Python (Perl developer back in the day..)  but think the above ought to delay only on new connection.

**Reply by canton7** [reply]
Aha, yep that's good

**Reply by paultweedy77** [reply]
I've upped to the latest release (v1.5.0) but patched in that sleep(1) hack. No adverse affects so far. 
Oddly enough the only time my inverter crashes is when I occasionally change something in HA, eg add a new sensor to a dashboard card. Not consistently, but when it does crash, it's often when I've done that.
I also noticed in the integration settings the max_read was set to 8 for some reason. I've now reset that to the default for direct LAN (100) which is resulting in far fewer TCP packets being sent to the inverter on each poll, which might help stability also.

**Reply by canton7** [reply]
Hmm. I can imagine a change in HA causing it to restart this integration, which means it establishes a new tcp connection, which is the point that the delay is supposed to help with! You can watch the log to see if that's what's actually happening.

Weird with the mac_read thing, I've tested that case many times. I'll try and repro it again...

**Reply by paultweedy77** [reply]
Quick update - more stable, HA changes don’t seem to crash the inverter, but the inverter is still crashing every 4-5 days for no obvious reason. 

Time for me to move to RS485 I think. Waveshare on order.

**Reply by jamesog** [reply]
I moved from LAN to RS485 last week (see #235) and haven't had a crash since.

**Reply by canton7** [reply]
> Quick update - more stable, HA changes don’t seem to crash the inverter, but the inverter is still crashing every 4-5 days for no obvious reason.

@paultweedy77 So your conclusion is that adding a delay there does help?

**Reply by paultweedy77** [reply, marked as answer]
Helps to stop HA changes crashing the inverter, yes. However random crashes do still happen (and don't happen when the integration is disabled). I have start the process to move to RS485 as a result (but that isn't going so well.. more in a new topic)

**Reply by canton7** [reply]
Cool thanks, I've included that in 1.5.3b5.
